### PR TITLE
[test] Demo 1

### PR DESCRIPTION
**Pull request by vladDotH**

### CODE REVIEW

**Comment by youngwow** [comment]
Чтобы добавить нейронку:

1. Создать файл, EfficientNetModel.py
2. Вставить следующий код:
`import tensorflow as tf
from tensorflow.keras.models import load_model
import tensorflow as tf
from sklearn.utils import class_weight


# Efficient net without augmentation4
class EfficientNetModel:
    metrics_names = ['loss', 'tp', 'fp', 'tn', 'fn', 'accuracy', 'precision', 'recall', 'auc', 'prc']
    METRICS = [
          tf.keras.metrics.TruePositives(name='tp'),
          tf.keras.metrics.FalsePositives(name='fp'),
          tf.keras.metrics.TrueNegatives(name='tn'),
          tf.keras.metrics.FalseNegatives(name='fn'), 
          tf.keras.metrics.BinaryAccuracy(name='accuracy'),
          tf.keras.metrics.Precision(name='precision'),
          tf.keras.metrics.Recall(name='recall'),
          tf.keras.metrics.AUC(name='auc'),
          tf.keras.metrics.AUC(name='prc', curve='PR'), # precision-recall curve
    ]
    
    def __init__(self, input_shape, num_classes):
        self.input_shape = input_shape
        self.num_classes = num_classes
        self.base_model = tf.keras.applications.efficientnet.EfficientNetB2(input_shape=self.input_shape, include_top=False, weights='imagenet')
        self.data_augmentation = tf.keras.Sequential([
            tf.keras.layers.RandomFlip("horizontal_and_vertical"),
            tf.keras.layers.RandomContrast(0.2),
            tf.keras.layers.RandomZoom(0.2),
            tf.keras.layers.RandomRotation(factor=0.15),
            tf.keras.layers.RandomTranslation(height_factor=0.1, width_factor=0.1),
            tf.keras.layers.RandomContrast(factor=0.1),
            ],
            name="img_augmentation",
            )
        self.model = self.build_model()
        
    def build_model(self):
        inputs = tf.keras.layers.Input(shape=self.input_shape)
        x = self.data_augmentation(inputs)
        x = self.base_model(x)
        
        for layer in self.base_model.layers[:40]:  #-20
            # if not isinstance(layer, tf.keras.layers.BatchNormalization):
            layer.trainable = False
                
        x = tf.keras.layers.GlobalAveragePooling2D()(x)
        predictions = tf.keras.layers.Dense(self.num_classes, activation='sigmoid')(x)  # !!
        model = tf.keras.models.Model(inputs=inputs, outputs=predictions)
        return model

     
    def compile(self, learning_rate = 0.0001):
        self.model.compile(optimizer=tf.keras.optimizers.Adam(
            learning_rate=learning_rate),
            loss='binary_crossentropy',
            metrics=self.METRICS)  
    
    def evaluate(self, data):
        return self.model.evaluate(data)
    
    def predict(self, data, verbose=None):
        return self.model.predict(data, verbose)
    
    def summary(self):
        return self.model.summary()
    
    def save_model(self, filepath):
        self.model.save(filepath)
        
    def load_model(self, model_path):
        self.model.load_weights(model_path)
        
    def plot_model_architecture(self, file_path):
        tf.keras.utils.plot_model(self, to_file=file_path, show_shapes=True)        
`
3. При использовании загрузить модель:
`model = EfficientNetModel(input_shape=(64, 64, 3), num_classes=1)
model.load_model('efficientB2_model.h5')
# define the input image size
input_size = (64, 64)`
Веса взять из ветки research: https://github.com/moevm/MSE-2023-image-ecology-12/tree/research/search/models/class_weight_augmentation/kaggle/working/weights

4. Создать функцию find_forest в папке worker/image_processing/find_forest/ и вставить следующий код:
`tf.keras.utils.disable_interactive_logging()

# тут img RGB снимок GeoTiff (img = get_image_RGB(filename)), 
def find_forest(img, block_size=64):
    threshold = 0.25

    binary_image = otsu_method(img)
    #show_image(binary_image)
    
    # loop through each block and save as a separate JPEG file
    for i in range(0, img.shape[1], block_size):
        left_block_predicted_label = 0
        for j in range(0, img.shape[0], block_size):
            # calculate the pixel coordinates of the block
            ulx = i
            uly = j
            lrx = i + block_size
            lry = j + block_size

            # make sure the block is within the image bounds
            if lrx > img.shape[0]:
                lrx = img.shape[0]
                ulx = lrx - block_size
            if lry > img.shape[1]:
                lry = img.shape[1]
                uly = lry - block_size

            block = img[ulx:lrx, uly:lry, :]

            # show_image(block)

            image = cv2.resize(block, input_size)
            image = image.astype('float32') / 255.0 

            prediction = model.predict(np.expand_dims(image, axis = 0))
            predicted_label = 1 if prediction.item() >= threshold else 0
            #print(predicted_label)

            if predicted_label == 0:
                #print(binary_image[ulx:lrx, uly:lry])
                #print()
                if ulx-64 >= 0 and lrx-64 >= 0:
                  if left_block_predicted_label == 1:
                      block = img[ulx-32:lrx-32, uly:lry, :]

                      image = cv2.resize(block, input_size)
                      image = image.astype('float32') / 255.0 

                      prediction = model.predict(np.expand_dims(image, axis = 0))
                      predicted_label = 1 if prediction.item() >= threshold - 0.2 else 0
                
                if predicted_label == 0:
                    binary_image[ulx:lrx, uly:lry] = 0

                """сравнивается, если 1 - 0 
                Добавить сравнение для 0 - 1, а также вертикали """
                #print(binary_image[ulx:lrx, uly:lry])

            # if predicted_label == 1:
            #     #print(binary_image[ulx:lrx, uly:lry])
            #     #print()
            #     if ulx-64 >= 0 and lrx-64 >= 0:
            #       if left_block_predicted_label == 0:
            #           block = img[ulx-32:lrx-32, uly:lry, :]

            #           image = cv2.resize(block, input_size)
            #           image = image.astype('float32') / 255.0 

            #           prediction = model.predict(np.expand_dims(image, axis = 0))
            #           predicted_label = 1 if prediction.item() >= 0.25 else 0
            #     if predicted_label == 0:
            #         binary_image[ulx:lrx, uly:lry] = 0
                #print(binary_image[ulx:lrx, uly:lry])
            left_block_predicted_label = predicted_label

    #show_image(binary_image)
    img_with_contours = draw_contours(binary_image, img)
    

    return img_with_contours`
    
 5. Заменить` for line in otsu_method(image_RGB):` на `for line in find_forest(image_RGB):` (в worker/app)

**Review by Saprigenie — APPROVED** [review]

**Review by dart-mih — APPROVED** [review]

**Review by dart-mih — APPROVED** [review]

**Review by Saprigenie — APPROVED** [review]
